### PR TITLE
[ fix/maintainance__hyphens-edge-cases ]

### PR DESCRIPTION
**Pull request by LaurelineP**
- if more than one hyphen in dir. name: will get the real
  - authorName *(based on found github url using curl)*
  - projectName: being the remaining string
  
Current implementation check the username in reverse direction:
Usually an existing username is appending a value to be unique
Ex: SJ ( author in card / link in card ) is not the same person as SJ-209
- https://github.com/SJ : is not the one that contributed
- https://github.com/SJ-209 : is the one that contributed - which we can retrieve in 
the fix implementation ( pict 1 ) 
  
### The Fix detecting folders name with multiple hyphens
image

### Former generated cards authorName and projectName deduction
image

image
image

### CODE REVIEW

**Comment by stilt0n** [comment]
This is an interesting idea. I hope you don't mind if I propose a possible alternative.

I think instead of checking to see if GitHub usernames exist for different hyphen combinations it would be a little more reliable to get the authors of the pull request and check for matches.

This is slightly tricky because we're running this script on `push` events so the pr isn't part of the job environment.

There's a few ways to do this and they're all honestly a little inconvenient. Easiest is probably to install github's [octokit](https://github.com/octokit/octokit.js) as a dev dependency with npm. 

I think then you can get most of what you need from the environment:

```js
const { Octokit } = require('@octokit/rest');

// returns all users associated with the commit being merged
const getPossibleContributors = async () => {
  const [owner, repo] = process.env.GITHUB_REPOSITORY.split('/');
  const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN });
  const result = await octokit.rest.repos.listPullRequestsAssociatedWithCommit({
    owner,
    repo,
    commit_sha: process.env.GITHUB_SHA,
  });
  return result.data.map(({ user }) => user.login);
};
```
There are docs for this [here](https://octokit.github.io/rest.js/v17/) but the are honestly not great.

Then you can use the list of users associated with the change to parse the directory:

```js
const candidates = await getPossibleContributors();
const authorName = candidates.find(candidate => dir.startsWith(candidate));
const projectName = dir.replace(authorName, '').slice(1);
```

**Comment by LaurelineP** [comment]
> This is an interesting idea. I hope you don't mind if I propose a possible alternative.
> 
> I think instead of checking to see if GitHub usernames exist for different hyphen combinations it would be a little more reliable to get the authors of the pull request and check for matches.
> 
> This is slightly tricky because we're running this script on `push` events so the pr isn't part of the job environment.
> 
> There's a few ways to do this and they're all honestly a little inconvenient. Easiest is probably to install github's [octokit](https://github.com/octokit/octokit.js) as a dev dependency with npm.
> 
> I think then you can get most of what you need from the environment:
> 
> ```js
> const { Octokit } = require('@octokit/rest');
> 
> // returns all users associated with the commit being merged
> const getPossibleContributors = async () => {
>   const [owner, repo] = process.env.GITHUB_REPOSITORY.split('/');
>   const octokit = new Octokit({ auth: process.env.GITHUB_TOKEN });
>   const result = await octokit.rest.repos.listPullRequestsAssociatedWithCommit({
>     owner,
>     repo,
>     commit_sha: process.env.GITHUB_SHA,
>   });
>   return result.data.map(({ user }) => user.login);
> };
> ```
> 
> There are docs for this [here](https://octokit.github.io/rest.js/v17/) but the are honestly not great.
> 
> Then you can use the list of users associated with the change to parse the directory:
> 
> ```js
> const candidates = await getPossibleContributors();
> const authorName = candidates.find(candidate => dir.startsWith(candidate));
> const projectName = dir.replace(authorName, '').slice(1);
> ```
Aloha,
@stilt0n Nice and interesting proposition !

I've proposed this approach (PR) so it does not add extra packages or installations.
We used to avoid adding extra tooling and or complexity when we could do otherwise for the ZTM PRs.


I let @MattCSmith provides his POV on wether we add packages or we remain without  - 
and merge the MR - I cannot merge my own PR :) 

Thanks again

**Comment by stilt0n** [comment]
That makes sense. I added the dependencies we were already using in the workflows to the package.json as dev dependencies so it might be good to know if I should change that back.

You can also use this approach without adding a dependency copying what [Wolfgang did here](https://github.com/zero-to-mastery/ZTM-Quest/blob/ae3b2495ee8aca2d41f9d2cb22d447a66d669400/.github/workflows/hacktoberfest-label.yaml#L22)

But this will require you to pass the output of that job into the script which would add a lot of complexity. I could see why you might want to avoid that.

**Comment by MattCSmith** [comment]
@LaurelineP @stilt0n 
Thank you both for taking a look into solving this issue, I think you both have very valid points and I don't think there is a perfect solution here. 

I think my main concern with checking if the Github Handle exists, is that this is likely to cause false positives too. 

Im still wondering at this point if we should just bring back having a meta.json with: 
{ github_handle: "mattcsmith", art_name: "Triangle" } inside each art directory. 

I know this adds back another files contributors have to add, but I think it would be far more accurate without the need for an actions to try and figure out the meta data stucture. 

What do you both think?

**Comment by stilt0n** [comment]
@MattCSmith 

Metadata sounds like the right direction to me. I think it *is* possible to get contributor information with high accuracy from pr information (would fail when the directory name is misspelled but metadata would too). But it's always worth questioning if it's worth adding a lot of extra complexity.

The cons to complexity are honestly pretty big:
- It's hard to test GitHub actions besides in production, and this are non-trivial changes
- Network calls have the potential to become a serious bottleneck because we repeat all old work on every merge
- We store "visited" metadata in `cards.json` but this adds additional job complexity

Meanwhile I honestly can't think of real cons to asking people to include a metadata file. Yes, it's a new file. But we were already asking for metadata in their directory names. If they include it in a file we can relax those naming conventions. It's hard to see this as a burden to contributors.

There is one (very minor) challenge with metadata which is how do we make it backwards compatible? But the obvious 
low effort way to do this is just check for metadata and use the directory name as a fallback. The main thing is that we can't 100% switch to metadata or we'll end up deleting the `cards.json` entries for all prior submissions.

tl;dr I am also leaning towards metadata

**Comment by LaurelineP** [comment]
Aloha @MattCSmith @stilt0n 
It's all good to have the metadata from users too.
This year project and revamp is very interesting: I liked the idea to avoid users ( here student ) to take more actions
and have the cards generated by github actions - it was really interesting to take advantages of github actions to run some js task.

But the prior versions of animation nations and user's implementation attempts - shown it could cause bottleneck on PRs & Review + asking students to update accordingly.

The metada approach would also requires reviewing existing one or may change the current dynamic in this crucial month - I remember last year we did not want to disturb students with extra students changes asked in the middle of hacktoberfest.
We could also plan for the next hactober to apply the changes if we have the same concerns as last year ( mentioned on the previous lines ) 

I would say it is up to the elements taken into account for this repo.
:)

**Comment by MattCSmith** [comment]
> Aloha @MattCSmith @stilt0n It's all good to have the metadata from users too. This year project and revamp is very interesting: I liked the idea to avoid users ( here student ) to take more actions and have the cards generated by github actions - it was really interesting to take advantages of github actions to run some js task.
> 
> But the prior versions of animation nations and user's implementation attempts - shown it could cause bottleneck on PRs & Review + asking students to update accordingly.
> 
> The metada approach would also requires reviewing existing one or may change the current dynamic in this crucial month - I remember last year we did not want to disturb students with extra students changes asked in the middle of hacktoberfest. We could also plan for the next hactober to apply the changes if we have the same concerns as last year ( mentioned on the previous lines )
> 
> I would say it is up to the elements taken into account for this repo. :)

I agree with making it as simple as possible for contributors. And I like that we are only asking for two files and one folder to be created. Its been an interesting experiment. But I think we have run into some issues with it, where there's not a perfect/accurate solution. 

Perhaps using the meta.json file does feel like a little bit of a step back, but I don't think we have used it in Animation Nation before, but it did seem to be a success in the other CSS project (CSS Art o Canvaz) we used it in. Since the Github Action will still populate the data for the includes.js. 

And I think the accuracy the meta.json bring warrants those few extra steps the contributor has to take and I do think that is the way we should proceed. However you raise a good point about making this change mid event.

**Comment by MattCSmith** [comment]
> There is one (very minor) challenge with metadata which is how do we make it backwards compatible? But the obvious low effort way to do this is just check for metadata and use the directory name as a fallback. The main thing is that we can't 100% switch to metadata or we'll end up deleting the `cards.json` entries for all prior submissions.
> 
@stilt0n Good point, I was thinking that a quick script could be written to iterate over the folders and create a meta.json in each of them. 

The script can then be deleted, the meta.json could be double checked. 
Shouldn't be too crazy.

**Comment by stilt0n** [comment]
I don't see the change midway through as hard solve. We'd update the docs and then we'd just keep the status quo as the backup:

```js
if (fs.existsSync(`${dir}/meta.json`)) {
  // push metadata from json file
} else {
  // get metadata from folder name
  // using already existing code
}
```
I also don't see this as much more of a burden on contributors.

The current process already requires contributors to add metadata to their directory name. The only difference is now it's in a json file.

And there's the added benefit that if you're a contributor and you notice your name is wrong you don't have to dig through how our CI works to fix it. You can just submit a PR that adds a meta.json file.
